java/se/sics/mspsim/chip/AT45DB.java

```diff
@@ -151,116 +151,105 @@ public void dataReceived(USARTSource source, int data) {
-          switch (data) {
-            case BUFFER1_TO_PAGE_ERASE, BUFFER2_TO_PAGE_ERASE -> {
-              if (DEBUG)
-                log("Buffer" + (data == BUFFER1_TO_PAGE_ERASE ? "1" : "2") + " to Page with Erase Command");
-              pos = 0;
-              setState(READ_ADDRESS);
-              next_state = data;
-              dummy = 0;
-              setReady(false);
-              source.byteReceived(0);
-            }
-            case BUFFER1_READ, BUFFER2_READ -> {
-              if (DEBUG)
-                log("Read Buffer Command " + (data == BUFFER1_READ ? "Buffer1" : "Buffer2"));
-              pos = 0;
-              setState(READ_ADDRESS);
-              next_state = data;
-              dummy = 1;
-              setReady(false);
-              source.byteReceived(0);
-            }
-            case BUFFER1_WRITE, BUFFER2_WRITE -> {
-              if (DEBUG)
-                log("Write Buffer Command " + (data == BUFFER1_WRITE ? "Buffer1" : "Buffer2"));
-              pos = 0;
-              setState(READ_ADDRESS);
-              next_state = data;
-              dummy = 0;
-              setReady(false);
-              source.byteReceived(0);
-            }
-            case PAGE_TO_BUFFER1, PAGE_TO_BUFFER2 -> {
+        switch (state) {
+          case READ_ADDRESS -> {
+            pos++;
+            if (pos == 1) {
+              pageAddress = ((data & 0xF) << 7);
+            } else if (pos == 2) {
+              pageAddress |= ((data & 0xFE) >> 1);
+              bufferAddress = ((data & 1) << 9);
+            } else if (pos == 3) {
+              bufferAddress |= data;
+
               if (DEBUG)
-                log("Page To Buffer " + (data == PAGE_TO_BUFFER1 ? "1" : "2") + " Command");
-              pos = 0;
-              setState(READ_ADDRESS);
-              next_state = data;
-              dummy = 0;
-              setReady(false);
-              source.byteReceived(0);
-            }
-            case STATUS_REGISTER_READ -> {
-              if (DEBUG) log("Read status register command.  status: " + status);
-              setState(STATUS_REGISTER_READ);
-              source.byteReceived(0);
+                log("Address - PA[10-0]: " + pageAddress + " BA[8-0]: " + bufferAddress);
+
+              if (dummy == 0) {
+                if (DEBUG) log("State " + state + " -> " + next_state);
+                setState(next_state);
+              }
+            } else {
+              if (--dummy == 0) {
+                if (DEBUG) log("State " + state + " -> " + next_state);
+                setState(next_state);
+              }
             }
-            default -> {
-              logw(WarningType.EMULATION_ERROR, "WARNING: Command not implemented: " + data);
-              source.byteReceived(0);
+            source.byteReceived(0);
+          }
+          case BUFFER1_READ, BUFFER2_READ -> {
+            // Return bytes from the RAM buffer
+            buf_num = (state == BUFFER1_READ ? 1 : 2);
```

Comment:
Github makes it hard to follow what happened here, but I'm guessing all lines are re-indented, and then Github decided to snap the diff view on a different place.

You might as well remove the parenthesis in the RHS here if the git blame gets trampled anyways.